app/models/custom_field/calculated_value.rb

```diff
@@ -73,29 +66,92 @@ def formula_string
+      end
+    end
+
     private
 
+    def calculator
+      Dentaku::Calculator.new(case_sensitive: true)
+    end
+
+    def valid_formula_syntax?
+      # Attempt to parse the formula. If no error is returned, the formula is syntactically valid.
+      calculator.ast(formula_str_without_patterns)
+      true
+    rescue Dentaku::ParseError, Dentaku::TokenizerError
+      false
+    end
+
     def formula_contains_only_allowed_characters?
       # List of allowed characters in a formula. This only performs a very basic validation.
-      # Allowed characters are:
-      # + - / * ( ) whitespace digits and decimal points
-      # Additionally, the formula may contain references to custom fields in the form of `cf_123` where 123 is the ID of
-      # the custom field.
+      # The allowed characters are:
+      # Our mathematical operators, whitespace, digits and decimal points
+      # Additionally, the formula may contain references to custom fields in the form of `{{cf_123}}`
+      # where 123 is the ID of the custom field.
       # Once this basic validation passes, the formula will be parsed and validated by Dentaku, which builds an AST
       # and ensures that the formula is really valid. A welcome side effect of the basic validation done here is that
       # it prevents built-in functions from being used in the formula, which we do not want to allow.
-      allowed_chars = %w[+ - / * ( )] + [" "]
-      allowed_tokens = /\A(cf_\d+|\d+\.?\d*|\.\d+)\z/
+      allowed_chars = MATH_OPERATORS_FOR_FORMULA + [" "]
+      allowed_tokens = /\A(\{\{cf_\d+}}|[\d.]+)\z/
 
       formula_string.split(Regexp.union(allowed_chars)).reject(&:empty?).all? do |token|
         token.match?(allowed_tokens)
       end
     end
 
     # Returns a list of custom field IDs used in the formula.
-    # For a formula like `2 + cf_12 + cf_4` it returns `[12, 4]`.
+    # For a formula like `2 + {{cf_12}} + {{cf_4}}` it returns `[12, 4]`.
     def cf_ids_used_in_formula(formula_str)
-      formula_str.scan(/\bcf_(\d+)\b/).flatten.map(&:to_i)
+      formula_str.scan(/(?<=\{\{cf_)\d+(?=}})/).map(&:to_i)
+    end
+
+    # Returns `formula_string` with all custom field references detokenized so that they are parseable by Dentaku.
+    # For example, for `2 + {{cf_12}} + {{cf_4}}` it returns `2 + cf_12 + cf_4`.
+    def formula_str_without_patterns
+      formula_string.gsub(/\{\{(cf_\d+)}}/, '\1')
+    end
+
+    def has_circular_reference?(custom_field, original_id, visited = Set.new)
+      return true if visited.include?(custom_field.id)
+
+      visited.add(custom_field.id)
+
+      if custom_field.field_format_calculated_value? && custom_field.formula
+        referenced_custom_fields = custom_field.formula.fetch("referenced_custom_fields", [])
+
+        return true if referenced_custom_fields.include?(original_id)
+
+        referenced_custom_fields.any? do |ref_id|
+          referenced_field = ProjectCustomField.find_by(id: ref_id)
```

Comment:
Probably more efficient to query all referenced in one go:
```suggestion
        ProjectCustomField.where(id: referenced_custom_fields).any? do |referenced_field|
```